### ISSUE TITLE
Maven publishing should allow retrying on connection timeout

### ISSUE THREAD

**Issue by anuraaga** [gradle/gradle#15360]
### Expected Behavior
Maven publishing retries, possibly as an opt-in flag, when connection times out to the remote.

### Current Behavior
Maven publishing fails if connection times out

### Context

We publish to bintray in a massive build with dozens of projects.

https://github.com/open-telemetry/opentelemetry-java-instrumentation

The connection can be flaky, with "connection timed out" errors. This build scan illustrates the failures

https://scans.gradle.com/s/f7dwe46xtdv5s/failure

A connection timed out is safe to retry, in the case that the repository is actually available at all, because nothing is published yet. If Gradle didn't automatically do so, if there was at least an option to retry on connection timed out, it can make this sort of publishing more stable.

**Comment by tprochazka** [comment]
I fully agree with that. It is an especially big issue if you have multiple modules and you want to deploy all of them. When one of them fails, you need to delete partial publish from the maven repo and repeat the process.

I think that in the ideal world there should be a special protocol for publishing to the repo which will support publish in transaction mode, but it is a different story ;-)

**Comment by rivancic** [comment]
I have the same issue.

I have project with many subprojects/sub-subprojects (~100). Project is being build on Jenkins then published to Maven repository. The more projects are introduced the higher chance that the gradle publish fails because of connection timeout. Its real issue. Will have to migrate to build in Jenkins publishing plugin as we see its stable. 

Big downside is the necessary complex configuration part for publishing with Jenkins plugin, that is supported out of the box with default config in gradle maven-publish plugin.

**Comment by ljacomet** [comment]
Sorry for the late reply!

Gradle does [retry on failure](https://github.com/gradle/gradle/pull/9445) since 5.6. You can even see that in the build scan [here](https://scans.gradle.com/s/f7dwe46xtdv5s/failure?expanded-stacktrace=WyIwLTEzLTI2LTM5LTQ1Il0&focused-stack-frame=0-13-26-39-45-30#1).
Support for [retrying on metadata download](https://github.com/gradle/gradle/issues/27262) was added in 8.7.

That later issue also contains information about internal properties to control the backoff and timeout.

You would need to investigate why the defaults were not enough for your case.
